lib/spack/spack/concretize.py

```diff
@@ -203,46 +202,101 @@ def concretize_version(self, spec):
 
         return True   # Things changed
 
-    def concretize_architecture(self, spec):
-        """If the spec is empty provide the defaults of the platform. If the
-        architecture is not a string type, then check if either the platform,
-        target or operating system are concretized. If any of the fields are
-        changed then return True. If everything is concretized (i.e the
-        architecture attribute is a namedtuple of classes) then return False.
-        If the target is a string type, then convert the string into a
-        concretized architecture. If it has no architecture and the root of the
-        DAG has an architecture, then use the root otherwise use the defaults
-        on the platform.
-        """
-        root_arch = spec.root.architecture
-        sys_arch = spack.spec.ArchSpec(spack.architecture.sys_type())
-        spec_changed = False
+    def _concretize_arch_component(self, component, platform, spec):
+        """ Concretizes the specified component of the architecture in a
+        spec.
+
+        On heterogeneous architectures it will look and see if the first
+        component will match the second component ie if the os is for the
+        front end then it will be sure to match the target with the front end.
+        On architectures with one single architecture it should work the
+        since both front end and back end are the same.
+
+        Args:
+            component (str):      which component to check and set in arch spec
+            platform  (Platform): platform object to help set the attributes of
+                                  architecture
+            spec      (Spec):     spec object to set the architecture on.
+
+        Returns:
+            Returns True if any of the components were set. If all components
+            are present then return False. True represents a change happened
+            and False represents no change."""
+        architecture = spec.architecture
+        #  represents whatever frontend exists, same if on single arch
+        front_end = "fe"
+
+        #  methods available to the platform class depends on component
+        #  so created a mapping between component and method name.
+        methods = {"platform_os": "operating_system", "target": "target"}
```

Comment:
I'd prefer an approach that  avoided the use of getattr. I'm thinking of something like the following:

```
frontend_os = platform.operating_system("fe")
frontend_target = platform.target("fe")
if architecture.target == frontend_target:
  architecture.operating_system = frontend_os
if architecture.operating_system == frontend_os:
  architecture.target = frontend_target
```

To be clear, this would omit the need for `component` and you'd only need 1 call to `_concretize_arch_component`.

I think this achieves the same result and is clearer. That being said one issue with it is that it will override the os/target even if they are already set (which I think your code does too).

Reply:
Was overthinking it here. I guess I would also need to set the backend appropriately.

Reply:
> I guess I would also need to set the backend appropriately.

Do you mean there is an analogous problem that when the user specifies `os=be` then you'd want to ensure that `target` is set to a backend target as well? If so I agree and that would be awesome of you to add, but IMO that's not required (adding a TODO would be helpful though).